sdk/log/batch_test.go

```diff
@@ -4,15 +4,169 @@
+	e := new(exporter)
+	b := NewBatchingProcessor(e)
+	_ = b.ForceFlush(context.Background())
+	require.True(t, e.forceFlushCalled, "exporter ForceFlush not called")
+}
+
+func TestBatchingProcessorShutdownCanceled(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	b := NewBatchingProcessor(slowExporter{})
+	err := b.Shutdown(ctx)
+	require.ErrorIs(t, err, context.Canceled)
+}
+
+func TestBatchingProcessorForceFlushCancled(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	b := NewBatchingProcessor(slowExporter{})
+	err := b.ForceFlush(ctx)
+	require.ErrorIs(t, err, context.Canceled)
+}
+
+func TestBatchingProcessorOnEmit(t *testing.T) {
+	e := new(exporter)
+	b := NewBatchingProcessor(e)
+	t.Cleanup(func() {
+		assert.NoError(t, b.Shutdown(context.Background()))
+	})
+
+	var r Record
+	r.SetSeverityText("test")
+	_ = b.OnEmit(context.Background(), r)
+	require.False(t, e.exportCalled, "exporter Export called")
+}
+
+func TestBatchingProcessorOnEmitForceFlush(t *testing.T) {
+	e := new(exporter)
+	b := NewBatchingProcessor(e)
+	t.Cleanup(func() {
+		assert.NoError(t, b.Shutdown(context.Background()))
+	})
+
+	var r Record
+	r.SetSeverityText("test")
+	_ = b.OnEmit(context.Background(), r)
+	_ = b.ForceFlush(context.Background())
+	require.True(t, e.exportCalled, "exporter Export not called")
+	assert.Equal(t, []Record{r}, e.records)
+}
+
+func TestBatchingProcessorOnEmitTick(t *testing.T) {
+	e := new(syncExporter)
+	s := NewBatchingProcessor(e, WithExportInterval(time.Millisecond))
+	t.Cleanup(func() {
+		assert.NoError(t, s.Shutdown(context.Background()))
+	})
+
+	var r Record
+	r.SetSeverityText("test")
+	_ = s.OnEmit(context.Background(), r)
+
+	require.Eventually(t, func() bool { return e.exportCalled.Load() }, time.Second, time.Millisecond, "exporter Export not called")
+	assert.Equal(t, []Record{r}, e.Records())
+}
+
+func TestBatchingProcessorOnFullQueue(t *testing.T) {
+	e := new(exporter)
+	b := NewBatchingProcessor(e, WithMaxQueueSize(1))
+	t.Cleanup(func() {
+		assert.NoError(t, b.Shutdown(context.Background()))
+	})
+
+	var r Record
+	r.SetSeverityText("test")
+	_ = b.OnEmit(context.Background(), r)
+	var r2 Record
+	r2.SetSeverityText("dropped")
+	_ = b.OnEmit(context.Background(), r2)
+
+	_ = b.ForceFlush(context.Background())
+	assert.Equal(t, []Record{r}, e.records)
+}
+
+func TestBatchingProcessorOnFullBatch(t *testing.T) {
+	e := new(exporter)
+	b := NewBatchingProcessor(e, WithExportMaxBatchSize(1))
+	t.Cleanup(func() {
+		assert.NoError(t, b.Shutdown(context.Background()))
+	})
+
+	var r Record
+	r.SetSeverityText("test")
+	_ = b.OnEmit(context.Background(), r)
+	var r2 Record
+	r2.SetSeverityText("on next export")
+	_ = b.OnEmit(context.Background(), r2)
+
+	_ = b.ForceFlush(context.Background())
+	assert.Equal(t, []Record{r}, e.records)
+
+	_ = b.ForceFlush(context.Background())
+	assert.Equal(t, []Record{r, r2}, e.records)
```

Comment:
A FaaS system is not going to continually call `ForceFlush` until there is nothing left to flush. This is the job of our processor to do. A test should be added that show all held records are flushed in a single call.